sycl/test-e2e/WorkGroupMemory/free_function_kernel.cpp

```diff
@@ -0,0 +1,83 @@
+// REQUIRES: aspect-usm_shared_allocations
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+#include <cassert>
+#include <sycl/detail/core.hpp>
+#include <sycl/ext/oneapi/experimental/work_group_memory.hpp>
+#include <sycl/ext/oneapi/free_function_queries.hpp>
+#include <sycl/group_barrier.hpp>
+#include <sycl/usm.hpp>
+
+using namespace sycl;
+
+queue q;
+context ctx = q.get_context();
+
+void sum_helper(sycl::ext::oneapi::experimental::work_group_memory<int[]> mem,
+                sycl::ext::oneapi::experimental::work_group_memory<int> ret,
+                size_t WGSIZE) {
+  ret = 0;
+  for (int i = 0; i < WGSIZE; ++i) {
+    ret += mem[i];
+  }
+}
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
+    (ext::oneapi::experimental::nd_range_kernel<1>))
+void sum(sycl::ext::oneapi::experimental::work_group_memory<int[]> mem,
+         int *buf, int *Result, size_t WGSIZE, bool UseHelper) {
+  const auto it = sycl::ext::oneapi::this_work_item::get_nd_item<1>();
+  size_t local_id = it.get_local_id();
+  mem[local_id] = buf[local_id];
+  group_barrier(it.get_group());
+  if (it.get_group().leader()) {
+    *Result = 0;
+    if (!UseHelper) {
+      for (int i = 0; i < WGSIZE; ++i) {
+        *Result += mem[i];
+      }
+    } else {
+      sycl::ext::oneapi::experimental::work_group_memory<int> ret;
+      sum_helper(mem, ret, WGSIZE);
+      *Result = ret;
+    }
+  }
+}
+
+void test(size_t SIZE, size_t WGSIZE, bool UseHelper) {
+  int *buf = malloc_shared<int>(WGSIZE, q);
+  assert(buf && "Shared USM allocation failed!");
+  int expected = 0;
+  for (int i = 0; i < WGSIZE; ++i) {
+    buf[i] = i;
+    expected += buf[i];
+  }
+  int *result = malloc_shared<int>(1, q);
+  assert(result && "Shared USM allocation failed!");
+#ifndef __SYCL_DEVICE_ONLY__
```

Comment:
It would be nice to add a comment saying that this `#ifndef` can be removed once the free function kernel limitations are fixed.  I think we are allowed to reference Jira numbers now, so we could add the Jira number to the comment.